packages/block-editor/src/components/block-media-update-progress/test/index.native.js

```diff
@@ -0,0 +1,501 @@
+	};
+	const callSaveCallback = ( payload ) => {
+		this.saveCallBack( payload );
+	};
+	const subscribeMediaUpload = ( callback ) => {
+		this.uploadCallBack = callback;
+	};
+	const subscribeMediaSave = ( callback ) => {
+		this.saveCallBack = callback;
+	};
+	const mediaSources = {
+		deviceCamera: 'DEVICE_CAMERA',
+		deviceLibrary: 'DEVICE_MEDIA_LIBRARY',
+		siteMediaLibrary: 'SITE_MEDIA_LIBRARY',
+	};
+	return {
+		subscribeMediaUpload,
+		subscribeMediaSave,
+		sendMediaUpload: callUploadCallback,
+		sendMediaSave: callSaveCallback,
+		mediaSources,
+	};
+} );
+
+// const MEDIA_ID_REMOTE = 123;
+const MEDIA_ID_LOCAL = 2;
+const MEDIA_ID_TEMP = 'tempid-0-1';
+
+const tempMediaFiles = [
+	{
+		alt: '',
+		caption: '',
+		id: 'tempid-0-1',
+		link: '',
+		mime: 'image/jpeg',
+		type: 'image',
+		url: '',
+	},
+	{
+		alt: '',
+		caption: '',
+		id: 'tempid-0-2',
+		link: '',
+		mime: 'image/jpeg',
+		type: 'image',
+		url: '',
+	},
+];
+
+const localMediaFiles = [
+	{
+		alt: '',
+		caption: '',
+		id: '2',
+		link: '',
+		mime: 'image/jpeg',
+		type: 'image',
+		url: '',
+	},
+	{
+		alt: '',
+		caption: '',
+		id: '3',
+		link: '',
+		mime: 'image/jpeg',
+		type: 'image',
+		url: '',
+	},
+];
+
+describe( 'BlockMediaUpdateProgress component', () => {
+	it( 'renders without crashing', () => {
+		const wrapper = shallow(
+			<BlockMediaUpdateProgress renderContent={ () => {} } />
+		);
+		expect( wrapper ).toBeTruthy();
+	} );
+
+	it( 'upload: listens media upload progress for local file', () => {
+		const progress = 10;
+		const payload = {
+			state: MEDIA_UPLOAD_STATE_UPLOADING,
+			mediaId: MEDIA_ID_LOCAL,
+			progress,
+		};
+
+		const onUpdateMediaUploadProgress = jest.fn();
+
+		const wrapper = shallow(
+			<BlockMediaUpdateProgress
+				onUpdateMediaUploadProgress={ onUpdateMediaUploadProgress }
+				mediaFiles={ localMediaFiles }
+				renderContent={ () => {} }
+			/>
+		);
+
+		sendMediaUpload( payload );
+
+		expect( wrapper.instance().state.progress ).toEqual( progress );
+		expect( wrapper.instance().state.isUploadInProgress ).toEqual( true );
+		expect( wrapper.instance().state.isUploadFailed ).toEqual( false );
+		expect( onUpdateMediaUploadProgress ).toHaveBeenCalledTimes( 1 );
+		expect( onUpdateMediaUploadProgress ).toHaveBeenCalledWith( payload );
+	} );
+
+	// UPLOAD tests
+	it( 'upload does not get affected by unrelated media uploads', () => {
+		const payload = {
+			state: MEDIA_UPLOAD_STATE_UPLOADING,
+			mediaId: 432, // id not belonging to assigned mediaFiles collection in test
+			progress: 20,
+		};
+		const onUpdateMediaUploadProgress = jest.fn();
+		const wrapper = shallow(
+			<BlockMediaUpdateProgress
+				onUpdateMediaUploadProgress={ onUpdateMediaUploadProgress }
+				mediaFiles={ localMediaFiles }
+				renderContent={ () => {} }
+			/>
+		);
+
+		sendMediaUpload( payload );
+
+		expect( wrapper.instance().state.progress ).toEqual( 0 );
+		expect( onUpdateMediaUploadProgress ).toHaveBeenCalledTimes( 0 );
```

Comment:
@mzorz I prefer to have each unit test having a single responsibility so I would do 

1) upload progress is not get affected by unrelated media uploads
2) onUpdateMediaUploadProgress is not called by unrelated media uploads. 

I think doing this makes the intent of the test clearer. Again, this is how I think of unit tests and I am open to discussing this further 😄

Reply:
:) 👍 I do understand where the suggestion is coming from, and I do agree one only point of failure per unit test sounds like a very good rule of thumb however I've been thinking about this and I don't think having _multiple checks_ necessarily goes against the abstraction of "unit testing" given when I think of it, it really depends on what the _unit_ we're testing is about, and the tools we have to express that the conditions for what we expect to happen are met. This last part, the tools, is an important one: the statements such as are `toHaveBeenCalledTimes` very fine grained, and are not enough to describe the actual meaning of the test. Having multiple checks shouldn't be confused with a signal that this is not unit testing. 

For example, the `toHaveBeenCalledTimes` method "unfortunately" only cares about the function call and not about the state change. Hence, it's impossible to check both the state change _and_ that such change has been made as a result of calling a certain method (and not through any other method call but that one) except by checking these 2 specific conditions have been met: the state change (with `expect( wrapper.instance().state.progress ).toEqual( 0 );`) and that such a state change (or lack of change) has been made by means of having called that specific method (or by the absence of such a call).

Let me know your thoughts!

Reply:
> :) 👍 I do understand where the suggestion is coming from, and I do agree one only point of failure per unit test sounds like a very good rule of thumb however I've been thinking about this and I don't think having multiple checks necessarily goes against the abstraction of "unit testing" given when I think of it, it really depends on what the unit we're testing is about, and the tools we have to express that the conditions for what we expect to happen are met. This last part, the tools, is an important one: the statements such as are toHaveBeenCalledTimes very fine grained, and are not enough to describe the actual meaning of the test. Having multiple checks shouldn't be confused with a signal that this is not unit testing.

This is a good point because if the group of assertions are focused on validating/verifying one core thing then it's still an unit. Thanks much for this explanation @mzorz Learning as I go along 😄 

> For example, the toHaveBeenCalledTimes method "unfortunately" only cares about the function call and not about the state change. Hence, it's impossible to check both the state change and that such change has been made as a result of calling a certain method (and not through any other method call but that one) except by checking these 2 specific conditions have been met: the state change (with expect( wrapper.instance().state.progress ).toEqual( 0 );) and that such a state change (or lack of change) has been made by means of having called that specific method (or by the absence of such a call).

> Let me know your thoughts!

Indeed. I understand 🙏